crates/datafusion_ext/src/metrics.rs

```diff
@@ -0,0 +1,252 @@
+/// skipped and metrics collection should be added to the execution plan
+/// directly.
+#[derive(Debug, Clone)]
+pub struct DataSourceMetricsExecAdapter {
+    child: Arc<dyn ExecutionPlan>,
+    metrics: ExecutionPlanMetricsSet,
+}
+
+impl DataSourceMetricsExecAdapter {
+    pub fn new(plan: Arc<dyn ExecutionPlan>) -> Self {
+        Self {
+            child: plan,
+            metrics: ExecutionPlanMetricsSet::new(),
+        }
+    }
+}
+
+impl ExecutionPlan for DataSourceMetricsExecAdapter {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> Arc<Schema> {
+        self.child.schema()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.child.output_partitioning()
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        self.child.output_ordering()
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.child.clone()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(Self::new(children[0].clone())))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        let stream = self.child.execute(partition, context)?;
+        Ok(Box::pin(BoxedStreamAdapater::new(
+            stream,
+            partition,
+            &self.metrics,
+        )))
+    }
+
+    fn statistics(&self) -> Statistics {
+        self.child.statistics()
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+}
+
+impl DisplayAs for DataSourceMetricsExecAdapter {
+    fn fmt_as(&self, _t: DisplayFormatType, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "DataSourceMetricsExecAdapter")
+    }
+}
+
+struct BoxedStreamAdapater {
+    stream: SendableRecordBatchStream,
+    metrics: DataSourceMetrics,
+}
+
+impl BoxedStreamAdapater {
+    fn new(
+        stream: SendableRecordBatchStream,
+        partition: usize,
+        metrics: &ExecutionPlanMetricsSet,
+    ) -> Self {
+        Self {
+            stream,
+            metrics: DataSourceMetrics::new(partition, metrics),
+        }
+    }
+}
+
+impl Stream for BoxedStreamAdapater {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let poll = self.stream.poll_next_unpin(cx);
+        self.metrics.record_poll(poll)
+    }
+}
+
+impl RecordBatchStream for BoxedStreamAdapater {
+    fn schema(&self) -> SchemaRef {
+        self.stream.schema()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct AggregatedMetrics {
+    /// Total time taken for a plan to execute.
+    pub elapsed_compute_ns: u64,
+    /// Total output rows for the plan.
+    pub output_rows: u64,
+    /// Total bytes processed.
+    pub bytes_processed: u64,
+}
```

Comment:
Not currently used, but I plan to make use of this when replacing the existing `session_query_metrics` table.